### PR TITLE
[model, ci] fix: migrate qwen3_5 linear-attn cache to transformers 5.9 layered DynamicCache

### PR DESCRIPTION
**Pull request by TimYangst**
### What does this PR do?

> Fixes a transformers 5.x regression in the `qwen3_5` / `qwen3_5_moe` gated-DeltaNet
> (linear-attention) cache path, surfaced after the 5.2 → 5.9 model upgrade, and adds a
> general `use_cache=True` test that closes the CI blind spot that let it through.
>
> transformers 5.x replaced the bespoke per-model `DynamicCache` (dict-like
> `conv_states[idx]` / `recurrent_states[idx]`, `has_previous_state` attribute) with the
> generic **layered** `DynamicCache`. The patch configs still used the old API, so any
> `use_cache=True` forward crashed — most notably `_update_linear_attn_mask`, whose
> upstream-verbatim caller now passes a `DynamicCache` where the override expected a
> tensor and called `.shape[-1]` on it.

### Checklist Before Starting

- Related: follow-up to the transformers 5.9 bump (#774).
- PR title follows `[{modules}] {type}: {description}` format.

### Test

> - New `tests/models/test_models_cache_forward_v5.py`: one `use_cache=True` prefill across
>   all 13 v5 models, asserting the cache populates. The existing bitwise test
>   (`test_models_logits_equal_v5`) runs `use_cache=False` and is blind to cache-API drift —
>   this is the gap that let the bug through.
> - Linear cases (`qwen3_5`, `qwen3_5_moe`) keep the real `linear_attention` layers (no
>   full-attention override) so `update_conv_state` / `update_recurrent_state` /
>   `_update_linear_attn_mask` actually fire; they need `fla` + GPU and skip otherwise.
> - The smoke is isolated to the cache path: linear cases run all non-DeltaNet ops eager
>   (incl. MoE) so they depend only on `fla` + `flash_attn` (the skip gates), not fused-MoE.
> - **13/13 pass on an 8xA100 worker** (transformers 5.9 + fla). Wired into
>   `gpu_unit_tests.yml` next to `test_models_logits_equal_v5`.
> - GPU-only by design (gated on `IS_CUDA_AVAILABLE`), matching the existing v5 forward-test
>   precedent; on NPU every case skips. `make quality` clean.

### API and Usage Example

> N/A — internal modeling / test change, no public API surface.

### Design & Code Changes

> - `qwen3_5_{gpu,npu}_patch_gen_config.py`: migrate the gated-DeltaNet cache to the v5
>   layered `DynamicCache`:
>   - `has_previous_state(layer_idx)` (method, not attribute);
>   - per-layer state via `cache_params.layers[idx].conv_states/.recurrent_states`, read only
>     under `use_precomputed_states` (the layered state is `None` until initialized);
>   - writes via `update_conv_state(state, idx)` / `update_recurrent_state(state, idx)`;
>   - `_update_linear_attn_mask` 2nd arg `cache_position` → `past_key_values`, detection via
>     `past_key_values.has_previous_state()` (host-side flag, still sync-free).
> - Regenerated `generated/` files (patchgen, drift-clean). `qwen3_5_moe` reuses the patch
>   functions via import, so its generated files update with no config diff of its own.
> - `tests/models/test_models_cache_forward_v5.py` + `gpu_unit_tests.yml` wiring.

### Checklist Before Submitting

- [x] Read the Contribute Guide
- [x] Applied pre-commit checks
- [x] Added/updated documentation (N/A — no API/config/workflow surface change)
- [x] Added tests to CI workflow

### CODE REVIEW

File: veomni/models/transformers/qwen3_5/qwen3_5_gpu_patch_gen_config.py [hunk 1]
```diff
@@ -323,14 +323,32 @@ def qwen3_5_gated_deltanet_forward_patched(
...
+    # disabled — it raises NotImplementedError — so Qwen3.5 here is training-only. When
+    # decode support is added, drop this term and let the host-side
+    # `has_previous_state(self.layer_idx)` flag decide on its own (matches upstream).
     use_precomputed_states = (
-        cache_params is not None and cache_params.has_previous_state and seq_len == 1 and cache_position is not None
+        cache_params is not None
+        and cache_params.has_previous_state(self.layer_idx)
+        and seq_len == 1
+        and cache_position is not None
     )
```
[the remark this answers is not in the document]
Comment (TimYangst) [reply veomni/models/transformers/qwen3_5/qwen3_5_gpu_patch_gen_config.py:346]: In transformers v5 the cache is contractually `Cache | None` (legacy tuple cache is gone), and upstream `modeling_qwen3_next` calls `has_previous_state` at the same sites without a `hasattr` guard. Adding one would diverge from upstream (patchgen drift) and silently mask a wrong-type cache instead of failing loudly. Skipping.

File: veomni/models/transformers/qwen3_5/qwen3_5_npu_patch_gen_config.py [hunk 2]
```diff
@@ -251,14 +251,30 @@ def qwen3_5_gated_deltanet_forward_patched(
...
+    # disabled — it raises NotImplementedError — so Qwen3.5 here is training-only. When
+    # decode support is added, drop this term and let the host-side
+    # `has_previous_state(self.layer_idx)` flag decide on its own (matches upstream).
     use_precomputed_states = (
-        cache_params is not None and cache_params.has_previous_state and seq_len == 1 and cache_position is not None
+        cache_params is not None
+        and cache_params.has_previous_state(self.layer_idx)
+        and seq_len == 1
+        and cache_position is not None
     )
```
[the remark this answers is not in the document]
Comment (TimYangst) [reply veomni/models/transformers/qwen3_5/qwen3_5_npu_patch_gen_config.py:272]: In transformers v5 the cache is contractually `Cache | None` (legacy tuple cache is gone), and upstream `modeling_qwen3_next` calls `has_previous_state` at the same sites without a `hasattr` guard. Adding one would diverge from upstream (patchgen drift) and silently mask a wrong-type cache instead of failing loudly. Skipping.

File: veomni/models/transformers/qwen3_5/qwen3_5_gpu_patch_gen_config.py [hunk 3]
```diff
@@ -532,20 +550,25 @@ def qwen3_5_text_model_update_linear_attn_mask(self, attention_mask, cache_posit
...
     """
     if attention_mask is None:
         return None
     # Cached forward (decode / continuation): see docstring — shapes wouldn't line up in
-    # apply_mask_to_padding_states, and upstream returns None here. Detected from shapes only.
-    if cache_position is not None and attention_mask.shape[-1] != cache_position.shape[-1]:
+    # apply_mask_to_padding_states, and upstream returns None here. `has_previous_state()` reads a
+    # host-side flag on the linear-attention cache layer (no 0-D GPU scalar), so this stays sync-free.
+    if past_key_values is not None and past_key_values.has_previous_state():
         return None
```
[the remark this answers is not in the document]
Comment (TimYangst) [reply veomni/models/transformers/qwen3_5/qwen3_5_gpu_patch_gen_config.py:572]: In transformers v5 the cache is contractually `Cache | None` (legacy tuple cache is gone), and upstream `modeling_qwen3_next` calls `has_previous_state` at the same sites without a `hasattr` guard. Adding one would diverge from upstream (patchgen drift) and silently mask a wrong-type cache instead of failing loudly. Skipping.

**Comment by Luosuu** [comment]
in what case, we will set use_cache=True?

**Comment by TimYangst** [comment]
> in what case, we will set use_cache=True?
Only in the generating path, basically it just a logic correctness fix.
